### "Solución MCO"

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

study = una matriz que contiene las horas de estudio. Esta es la variable explicativa. <br>
score = una matriz que contiene las puntuaciones de las pruebas. Esta es la variable de respuesta.

In [2]:
study = np.array([ 3, 4.5, 6, 1.2, 2, 6.9, 6.7, 5.5]).reshape(-1,1)
score = np.array([ 88, 85, 90, 80, 81, 92, 95, 90]).reshape(-1,1)

1). Defino la clase 'linearRegression' 

In [88]:
class linearRegression:
    
    def __init__(self, fit_intercept=True):
        self.coef_ = None
        self.intercept_ = None
        self._fit_intercept = fit_intercept
        self.r2_score = None
    
    def train(self, X, Y):
        n = X.shape[0]
        m = X.shape[1]
        intercept = np.ones((X.shape[0], 1))
        X1 = np.concatenate((intercept, X), axis = 1)
        B1 = np.dot(np.dot(np.linalg.inv(np.dot(X1.T, X1)), X1.T), Y) # beta
        self._fit_intercept = B1
        
        for i in range(len(self._fit_intercept)):
            self.intercept_ = B1[i-1]    # coeficiente constante
            self.coef_ = B1[i]           # coeficiente beta

        y_hat = self.intercept_ + X * self.coef_
        MC = np.corrcoef(Y.T, y_hat.T) # MC = matriz de correlación
        r = MC[0,1]                     # correlación
        r2_score = r**2    # r(yhat,Y)^2 = R^2
        self.r2_score = r2_score
        
    def output(self):
        df = pd.DataFrame(self._fit_intercept, index=['Intercept', 'X1'], columns=['Coefficient'])  # salida
        print(df)
        print('-'*30)
        print('La bondad de ajuste (r_2) es:', np.round(self.r2_score, 6))
        print('La constante (β0) es:', self.intercept_)
        print('La beta (β1) es:', self.coef_)

**Ejecución:**

In [89]:
LR = linearRegression()                    # Crea un objeto de regresión lineal.
LR.train(study, score)                     # Entrenamiento.
LR.output()                                # Muestra el resultado del entrenamiento.

           Coefficient
Intercept    77.684670
X1            2.221303
------------------------------
La bondad de ajuste (r_2) es: 0.845842
La constante (β0) es: [77.68466978]
La beta (β1) es: [2.22130284]


Y = β + β1 + u(residuo)

Puntuación = 77.68 + 2.22 Horas/estudio

La puntuación aumenta 2.22 puntos por cada hora de estudio.

Con un R^2 de 0,84 el modelo lineal obtenido tiene representatividad. El modelo explica.